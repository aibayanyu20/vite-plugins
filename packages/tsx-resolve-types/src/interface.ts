import type { ParseResult } from '@babel/parser'
import type { File } from '@babel/types'

export type AST = ParseResult<File>

export interface UserOptions {
  /**
   * @description 是否开启emits的类型推断
   * @default true
   */
  emits?: boolean
  /**
   * @description 是否开启props的类型推断
   * @default true
   */
  props?: boolean
}
