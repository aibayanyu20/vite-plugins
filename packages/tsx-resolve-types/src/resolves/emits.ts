import type { CallExpression, Expression, Identifier, ObjectExpression, Pattern, RestElement } from '@babel/types'
import {
  arrayExpression,
  identifier,
  isArrowFunctionExpression,
  isFunctionExpression,
  isIdentifier,
  isObjectExpression,
  isObjectMethod,
  isObjectPattern,
  isObjectProperty,
  isTSTypeReference,
  objectExpression,
  objectProperty,
  stringLiteral,
} from '@babel/types'
import { extractRuntimeEmits } from '@vue/compiler-sfc'
import type { CreateContextType } from '../utils/context'

function addEmitsType(exp: Identifier | RestElement | Pattern, ctx: CreateContextType) {
  if (isObjectPattern(exp) || isIdentifier(exp)) {
    const typeAnnotation = exp.typeAnnotation
    if (typeAnnotation && 'typeAnnotation' in typeAnnotation) {
      const node = typeAnnotation.typeAnnotation
      if (isTSTypeReference(node) && node.typeName && 'name' in node.typeName && node.typeName.name === 'SetupContext') {
        const typeParameters = node.typeParameters
        if (typeParameters && typeParameters.params) {
          const params = typeParameters.params
          const prop = params[0]
          if (prop)
            ctx.ctx.emitsTypeDecl = prop
        }
      }
    }
  }
}

function addEmitsToFunc(exp: CallExpression, ast: Expression) {
  const arg1 = exp.arguments[1]
  if (arg1) {
    if (isObjectExpression(arg1)) {
      const propsProperty = arg1.properties.find(p => isObjectProperty(p) && 'name' in p.key && p.key.name === 'emits')
      if (!propsProperty)
        arg1.properties.unshift(objectProperty(identifier('emits'), ast))
    }
  }
  else {
    const obj = objectExpression([objectProperty(identifier('emits'), ast)])
    exp.arguments.push(obj)
  }
}
function getSetupContext(expression: ObjectExpression, ctx: CreateContextType) {
  /**
   * 检索setup的函数
   */
  const properties = expression.properties
  const setup = properties.find((property) => {
    return property.type === 'ObjectMethod' && 'name' in property.key && property.key.name === 'setup'
  })

  /**
   * 存在setup的情况下
   */
  if (setup && isObjectMethod(setup)) {
    const params = setup.params
    const setupCtx = params[1]
    if (setupCtx) {
      // 处理添加数据
      addEmitsType(setupCtx, ctx)
    }
  }
}

function getEmitsAst(ctx: CreateContextType) {
  if (ctx.ctx.emitsTypeDecl) {
    // 开始执行
    const emits = extractRuntimeEmits(ctx.ctx)
    ctx.ctx.emitsTypeDecl = undefined
    if (emits && emits.size) {
      const keys = Array.from(emits).map((v) => {
        return stringLiteral(v)
      })

      return arrayExpression(keys)
    }
  }
}

export function resolveEmits(expression: CallExpression, ctx: CreateContextType) {
  /**
   * 拿到props的第一个参数
   */
  const argument = expression.arguments[0]
  if (!argument)
    return

  if (isFunctionExpression(argument) || isArrowFunctionExpression(argument)) {
    const params = argument.params
    if (params.length) {
      const emits = params[1]
      if (emits)
        addEmitsType(emits, ctx)
      const emitAst = getEmitsAst(ctx)
      if (emitAst)
        addEmitsToFunc(expression, emitAst)
    }
  }
  else if (isObjectExpression(argument)) {
    // 当前是Object的模式
    const emitsProp = argument.properties.find(p => isObjectProperty(p) && 'name' in p.key && p.key.name === 'emits')
    if (!emitsProp) {
      getSetupContext(argument, ctx)
      const emitAst = getEmitsAst(ctx)
      if (emitAst) {
        // 存在ast
        argument.properties.unshift(objectProperty(identifier('emits'), emitAst))
      }
    }
  }
}
