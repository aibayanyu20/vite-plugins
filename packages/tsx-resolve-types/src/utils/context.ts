import fs from 'node:fs'
import type { SimpleTypeResolveContext } from '@vue/compiler-sfc'
import type { AST } from '../interface'
import { createAst } from './ast'
import type { GraphContext } from './graphContext'

export interface CreateContextType {
  ctx: SimpleTypeResolveContext
  ast: AST
  source: string
  filepath: string
  importMergeDefaults?: boolean
}

export function createContext(code: string, id: string, graphCtx?: GraphContext): CreateContextType {
  const ast = createAst(code)
  const helper = new Set<string>()
  return {
    ast,
    filepath: id,
    source: code,
    importMergeDefaults: false,
    ctx: {
      filename: id,
      source: code,
      ast: ast.program.body,
      error(node, msg) {
        console.error(msg, node)
        throw new Error(node)
      },
      helper(key) {
        helper.add(key)
        return `_${key}`
      },
      getString(node) {
        return code.slice(node.start!, node.end!)
      },
      propsTypeDecl: undefined,
      propsRuntimeDefaults: undefined,
      propsDestructuredBindings: Object.create(null),
      emitsTypeDecl: undefined,
      isCE: false,
      options: {
        fs: {
          fileExists(_file: string): boolean {
            // 检查文件是否存在
            try {
              const stat = fs.statSync(_file)
              if (stat.isFile()) {
                if (!(/node_modules/.test(_file))) {
                  graphCtx?.add(_file, id)
                  graphCtx?.addGraph(id, _file)
                }

                return true
              }
              return false
            }
            catch (e) {
              return false
            }
          },
          readFile(_file: string): string | undefined {
            return fs.readFileSync(_file, 'utf-8')
          },
        },
      },
    },
  }
}
