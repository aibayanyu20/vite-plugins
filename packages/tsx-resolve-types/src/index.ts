import type { PluginOption } from 'vite'
import { GraphContext } from './utils/graphContext'
import { invalidateTypeCacheId } from './utils/invalidate'
import { getDepModules } from './utils/depModules'

export function tsxResolveTypes(): PluginOption {
  const graphCtx = new GraphContext()
  return {
    name: 'tsx-resolve-types',

    /**
     * 热更新
     */
    handleHotUpdate({ file, server, modules }) {
      if (graphCtx.has(file)) {
        // 清理缓存
        invalidateTypeCacheId(file)
        // 如果存在那么响应的所有的moduleGraph都需要更新
        const ids = graphCtx.get(file)
        const affectedModules = []
        if (ids) {
          for (const id of ids) {
            const mods = getDepModules(id, graphCtx, server.moduleGraph)
            graphCtx.removeCache(id)
            if (mods.length)
              affectedModules.push(...mods)
          }
          return [...modules, ...affectedModules]
        }
      }
      return modules
    },
  }
}
