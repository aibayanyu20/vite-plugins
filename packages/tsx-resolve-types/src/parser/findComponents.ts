import type { CallExpression } from '@babel/types'
import { isCallExpression } from '@babel/types'
import type { AST } from '../interface'
import { traverse } from '../utils/traverse'

function checkIsDefineComponent(node: CallExpression) {
  return isCallExpression(node) && node.callee && (node as any).callee?.name === 'defineComponent'
}
export function findComponents(ast: AST) {
  const componentsAst: CallExpression[] = []
  traverse(ast, {
    CallExpression({ node }) {
      if (checkIsDefineComponent(node))
        componentsAst.push(node)
    },
  })
  return componentsAst
}
