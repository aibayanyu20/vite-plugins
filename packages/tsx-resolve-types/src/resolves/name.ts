import type { AST } from '../interface'
import { traverse } from '../utils/traverse'

export function resolveName(ast: AST) {
  //  TODO: 从 ast 中解析出 name
  traverse(ast, {

  })
}
