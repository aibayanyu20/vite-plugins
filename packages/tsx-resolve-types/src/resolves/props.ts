import type {
  ArrowFunctionExpression,
  CallExpression,
  Expression,
  FunctionExpression,
  Identifier,
  ObjectExpression,
  Pattern,
  RestElement,
} from '@babel/types'
import {
  identifier,
  isArrowFunctionExpression,
  isAssignmentPattern,
  isFunctionExpression,
  isIdentifier,
  isObjectExpression,
  isObjectProperty,
  objectExpression,
  objectProperty,
} from '@babel/types'
import { extractRuntimeProps } from '@vue/compiler-sfc'
import { parseExpression } from '@babel/parser'
import type { CreateContextType } from '../utils/context'

function getTypeAnnotation(node: Identifier) {
  if (node.typeAnnotation && 'typeAnnotation' in node.typeAnnotation)
    return node.typeAnnotation
}

function addTypes(prop: RestElement | Identifier | Pattern | undefined, ctx: CreateContextType) {
  if (prop && isAssignmentPattern(prop)) {
    if (isIdentifier(prop.left) && !ctx.ctx.propsTypeDecl) {
      const type = getTypeAnnotation(prop.left)
      if (type)
        ctx.ctx.propsTypeDecl = type
    }
    if (prop.right)
      ctx.ctx.propsRuntimeDefaults = prop.right

    return prop.left
  }
  else if (prop && isIdentifier(prop)) {
    const type = getTypeAnnotation(prop)
    if (type)
      ctx.ctx.propsTypeDecl = type
  }
}
function getObjectSetup(expression: ObjectExpression, ctx: CreateContextType) {
  const properties = expression.properties
  const setup = properties.find((property) => {
    return property.type === 'ObjectMethod' && 'name' in property.key && property.key.name === 'setup'
  })
  if (setup?.type === 'ObjectMethod') {
    const params = setup.params
    const prop = params[0]
    const leftAst = addTypes(prop, ctx)
    if (leftAst)
      params[0] = leftAst as any
  }
}

function getPropsTypeToDefine(exp: CallExpression, ctx: CreateContextType) {
  /**
   * 直接在函数的泛型数据中获取
   */
  if (exp.typeParameters) {
    const params = exp.typeParameters?.params
    if (params) {
      const prop = params[0]
      if (prop)
        ctx.ctx.propsTypeDecl = prop
    }
  }
}

function getFuncType(exp: ArrowFunctionExpression | FunctionExpression, ctx: CreateContextType) {
  const params = exp.params
  const prop = params[0]
  const left = addTypes(prop, ctx)
  if (left)
    params[0] = left as any
}

function getPropsStr(ctx: CreateContextType) {
  if (ctx.ctx.propsTypeDecl) {
    // 开始执行函数获取里面的数据信息
    const propStr = extractRuntimeProps(ctx.ctx)
    ctx.ctx.propsTypeDecl = undefined
    ctx.ctx.propsRuntimeDefaults = undefined
    if (propStr) {
      /**
       * 判断生成的代码中是否存在
       */
      if (!ctx.importMergeDefaults)
        (propStr.includes('/*#__PURE__*/_mergeDefaults') || propStr.includes('/*@__PURE__*/_mergeDefaults') || propStr.includes('_mergeDefaults')) && (ctx.importMergeDefaults = true)
      return parseExpression(propStr)
    }
  }
}

function addPropsToFunc(exp: CallExpression, ast: Expression) {
  const arg1 = exp.arguments[1]
  if (arg1) {
    if (isObjectExpression(arg1)) {
      const propsProperty = arg1.properties.find(p => isObjectProperty(p) && 'name' in p.key && p.key.name === 'props')
      if (!propsProperty)
        arg1.properties.unshift(objectProperty(identifier('props'), ast))
    }
  }
  else {
    const obj = objectExpression([objectProperty(identifier('props'), ast)])
    exp.arguments.push(obj)
  }
}
export function resolveProps(expression: CallExpression, ctx: CreateContextType) {
  // 判断的第一个参数的类型
  const props = expression.arguments[0]
  if (!props)
    return
  getPropsTypeToDefine(expression, ctx)
  // 判断第一个参数的类型，是不是一个函数
  if (isFunctionExpression(props) || isArrowFunctionExpression(props)) {
    getFuncType(props, ctx)
    const propAst = getPropsStr(ctx)
    if (propAst)
      addPropsToFunc(expression, propAst)
  }
  else if (props.type === 'ObjectExpression') {
    // 当前是一个对象，就需要判断是否存在一个属性是props
    const propsProperty = props.properties.find(p => isObjectProperty(p) && 'name' in p.key && p.key.name === 'props')
    if (!propsProperty) {
      getObjectSetup(props, ctx)
      const propAst = getPropsStr(ctx)
      if (propAst)
        props.properties.unshift(objectProperty(identifier('props'), propAst))
    }
  }
}
