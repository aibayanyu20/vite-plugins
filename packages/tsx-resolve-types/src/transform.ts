import { createContext } from './utils/context'
import { findComponents } from './parser'
import { resolveProps } from './resolves/props'
import { resolveEmits } from './resolves/emits'
import { checkMergeDefaults } from './utils/checkMergeDefaults'
import type { GraphContext } from './utils/graphContext'
import type { UserOptions } from './interface'
import { generate } from './utils/genrate'

export function transform(code: string, id: string, graphCtx: GraphContext, options: UserOptions = {}) {
  if (options.props === false && options.emits === false)
    return code

  const ctx = createContext(code, id, graphCtx)
  const expression = findComponents(ctx.ast)
  for (const callExpression of expression) {
    if (options.props !== false)
      resolveProps(callExpression, ctx)
    if (options.emits !== false)
      resolveEmits(callExpression, ctx)
  }
  checkMergeDefaults(ctx)
  const gen = generate(ctx.ast)
  return {
    code: gen.code,
    map: gen.map,
  }
}
