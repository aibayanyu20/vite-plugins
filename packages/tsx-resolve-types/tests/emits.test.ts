import path from 'node:path'
import { describe, expect, it } from 'vitest'
import generate from '@babel/generator'
import { createContext } from '../src/utils/context'
import { findComponents } from '../src/parser'
import { resolveEmits } from '../src/resolves/emits'
import { basePath } from './fixtures/constant'
import inlineRaw from './fixtures/emits/inline.tsx?raw'
import funcRaw from './fixtures/emits/func.tsx?raw'

describe('emits', () => {
  it('inline', () => {
    const code = compiler(inlineRaw, 'inline.tsx')
    expect(code).toMatchSnapshot()
  })

  it('func', () => {
    const code = compiler(funcRaw, 'func.tsx')
    expect(code).toMatchSnapshot()
  })
})
function compiler(raw: string, id: string) {
  const ctx = createContext(raw, path.resolve(basePath, `emits/${id}`))
  const expression = findComponents(ctx.ast)
  for (const callExpression of expression)
    resolveEmits(callExpression, ctx)

  return generate(ctx.ast).code
}
