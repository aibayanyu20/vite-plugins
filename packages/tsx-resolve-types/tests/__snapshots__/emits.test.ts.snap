// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`emits > func 1`] = `
"import type { SetupContext } from 'vue';
import { defineComponent } from 'vue';
export const Func1 = defineComponent((_, ctx: SetupContext<{
  'change': [string];
}>) => {
  const func = () => {
    ctx.emit('change', 'func');
  };
  return () => {
    return <div></div>;
  };
}, {
  emits: [\\"change\\"]
});
interface Emits {
  'change': [string];
  'click': [Event];
}
export const Func2 = defineComponent((_, ctx: SetupContext<Emits>) => {
  const func = () => {
    ctx.emit('change', 'func');
  };
  return () => {
    return <div></div>;
  };
}, {
  emits: [\\"change\\", \\"click\\"]
});
export default defineComponent({
  emits: [\\"change\\", \\"click\\"],
  setup(_, ctx: SetupContext<Emits>) {
    return () => <div>func</div>;
  }
});
interface Emits1 {
  'click': [Event];
}
export const Func3 = defineComponent((_, {
  emit
}: SetupContext<Emits1>) => {
  return () => {
    return <div> Func</div>;
  };
}, {
  emits: [\\"click\\"]
});"
`;

exports[`emits > inline 1`] = `
"import type { SetupContext } from 'vue';
import { defineComponent } from 'vue';
interface Emits {
  'change': [string];
  'click': [Event];
}
export const Inline1 = defineComponent({
  emits: [\\"change\\"],
  setup(props, {
    emit
  }: SetupContext<{
    'change': [string];
  }>) {
    const func = () => {
      emit('change', 'inline');
    };
    return () => {
      return <div>
          Inline1
        </div>;
    };
  }
});
export const Inline2 = defineComponent({
  emits: [\\"change\\", \\"click\\"],
  setup(props, {
    emit
  }: SetupContext<Emits>) {
    const func = () => {
      emit('change', 'inline');
    };
    return () => {
      return <div>
          Inline2
        </div>;
    };
  }
});
export const Inline3 = defineComponent({
  emits: [\\"change\\", \\"click\\"],
  setup(props, ctx: SetupContext<Emits>) {
    const func = () => {
      ctx.emit('change', 'inline');
    };
    return () => {
      return <div>
          Inline2
        </div>;
    };
  }
});"
`;
